File: exercises/week5/series/sources-solution/series.py
from abc import ABC, abstractmethod


class Series(ABC):

    def __init__(self):
        self.current_term = 0
        self.current_value = 0

    def __del__(self):
        pass

    def compute(self, N):
        if self.current_term <= N:
            N -= self.current_term
        else:
            self.current_value = 0.
            self.current_term = 0

        for k in range(0, N):
            self.addTerm()

        return self.current_value

    @abstractmethod
    def getAnalyticPrediction(self):
        pass

    def addTerm(self):
        self.current_term += 1
        self.current_value += self.computeTerm(self.current_term)

    def computeTerm(self, k):
        pass

File: exercises/week6/series/sources-solution/src/series.cc
#include "series.hh"
#include <cmath>
/* -------------------------------------------------------------------------- */

Series::Series() {}

/* -------------------------------------------------------------------------- */

double Series::compute(UInt N) {

  if (current_term <= N)
    N -= this->current_term;
  else {
    this->current_value = 0.;
    this->current_term = 0;
  }

  for (UInt k = 0; k < N; ++k) {
    ++current_term;
    current_value += func(current_term);
  }

  return current_value;
}

/* -------------------------------------------------------------------------- */
double Series::getAnalyticPrediction() { return nan(""); }
